### ISSUE TITLE
File size increase in 6.4.3 when using dynamic_context

### ISSUE THREAD

**Issue by davidszotten** [coveragepy/coveragepy#1586]
**Describe the bug**
Hi. I suspect this is related to #1538, hence the similar name, but this is a different aspect. My symptom is a huge increase in coverage data file size (~4 -> 200Mb)

**To Reproduce**
I made a test project with 3 pytest files ,each with a single empty test, and added `dynamic_context = test_function` to a coveragerc

running pytest under coverage and then reading the line_bits table, i get
(`coverage run -m pytest tests/ ; sqlite3 .coverage 'select path, context from line_bits join file on file_id = file.id join context on context_id = context.id'`)

on 6.4.2
```
tests/test_1.py|
tests/test_1.py|test_1.test1
tests/test_2.py|
tests/test_2.py|test_2.test2
tests/test_3.py|
tests/test_3.py|test_3.test3
```

on 6.4.3
```
tests/test_1.py|
tests/test_1.py|test_1.test1
tests/test_1.py|test_2.test2
tests/test_1.py|test_3.test3
tests/test_2.py|
tests/test_2.py|test_1.test1
tests/test_2.py|test_2.test2
tests/test_2.py|test_3.test3
tests/test_3.py|
tests/test_3.py|test_1.test1
tests/test_3.py|test_2.test2
tests/test_3.py|test_3.test3
```

It looks like there is now a line for every file for every context.

Is this expected? Avoidable? The fix for #1538 seems to focus on queries. does that mean we can't avoid storing the extra data in the first place?

**Comment by nedbat** [comment]
Can you share the code for your test project?

**Comment by davidszotten** [comment]
sure. https://github.com/davidszotten/coverage-file-size

**Comment by davidszotten** [comment]
i might be missing a bunch of context/knowledge, but from what i can tell, while https://github.com/nedbat/coveragepy/commit/f54428fb8dfe98aacecc0b66e3c1ca2071ce1834 fixes the issue on top of 6.4.2 through 7.0.5, as of 7.10.0,  at least the [example from cython](https://github.com/cython/cython/issues/3515#issuecomment-1402586598) still passes after reverting f54428fb8d

**Comment by davidszotten** [comment]
was https://github.com/nedbat/coveragepy/commit/4cc32922685c6#diff-e5185ca855df54af4f94c983f99b52cfad9ddc17386353d09acdfa994dd72a33L574-L579 an alternative (and better) fix for the cython issue than adding lots of empty file-context entries?

**Comment by andersk** [comment]
@davidszotten My testing agrees, so we can revert f54428fb8dfe98aacecc0b66e3c1ca2071ce1834:

- #1629

**Comment by nedbat** [comment]
Fixed in 6df9ee56.

**Comment by nedbat** [comment]
This is now released as part of [coverage 7.2.7](https://pypi.org/project/coverage/7.2.7).

**Comment by davidszotten** [comment]
Thanks!
